React/Base/macOS/RCTUIKit.m

```diff
@@ -812,13 +814,21 @@ - (void)setImage:(UIImage *)image
   CALayer *layer = [self layer];
   
   if ([layer contents] != image || [layer backgroundColor] != nil) {
-    if (_tintColor != nil) {
-      image = [image copy];
-      [image lockFocus];
-      [_tintColor set];
-      NSRect imageRect = { NSZeroPoint, image.size };
-      NSRectFillUsingOperation(imageRect, NSCompositingOperationSourceIn);
-      [image unlockFocus];
+    if (_tintColor) {
+      if (!_tintingLayer) {
+        _tintingLayer = [CALayer new];
+        [_tintingLayer setFrame:self.bounds];
+        [_tintingLayer setAutoresizingMask:kCALayerWidthSizable | kCALayerHeightSizable];
+        [_tintingLayer setZPosition:1.0];
+        CIFilter *sourceInCompositingFilter = [CIFilter filterWithName:@"CISourceInCompositing"];
+        [sourceInCompositingFilter setDefaults];
+        [_tintingLayer setCompositingFilter:sourceInCompositingFilter];
+        [layer addSublayer:_tintingLayer];
+      }
+      [_tintingLayer setBackgroundColor:_tintColor.CGColor];
+    } else {
+      [_tintingLayer removeFromSuperlayer];
+      _tintingLayer = nil;
```

Comment:
Elsewhere, we set the filter directly on the layer. Should we do that here? Feels cleaner than adding and removing a sublayer.
https://github.com/microsoft/react-native-macos/blob/6659de8f425876ef7712d51c3e11f4b358de8364/React/Base/macOS/RCTUIKit.m#L727

An aside, I thought we had to override a combination of `wantsLayer/wantsUpdateLayer` to YES in order to properly modify the layer, but I don't see us doing that here. Maybe NSImageView is giving us some good defaults?

https://developer.apple.com/documentation/appkit/nsview/1483695-wantslayer
> If the [wantsUpdateLayer](https://developer.apple.com/documentation/appkit/nsview/1483461-wantsupdatelayer) method returns false, you should not interact with the underlying layer object directly. Instead, use the methods of this class to make any changes to the view and its layer. If wantsUpdateLayer returns true, it is acceptable (and appropriate) to modify the layer in the view’s [updateLayer()](https://developer.apple.com/documentation/appkit/nsview/1483580-updatelayer) method.

Reply:
More context: modifying the layer has caused issues before: https://github.com/microsoft/react-native-macos/issues/466

Reply:
RCTUIImageView is set up to be layer-hosting. A CALayer is being assigned to the `layer` property before setting `wantsLayer` to `YES`, which enables layer-hosting instead of layer-backing.

Since we don't do any rendering through `drawRect` there's no need for layer-backing. With layer-hosting we can do everything we want with the CALayer.

Reply:
Derp I just saw that we set it in `init`, never mind!

Reply:
How do you feel about setting the filter directly on the layer rather than creating a new sublayer? I suppose it may be an arbitrary choice, I mostly wanted to check that one way isn't better than the other in case we should update either usage.

Reply:
> Elsewhere, we set the filter directly on the layer. Should we do that here? Feels cleaner than adding and removing a sublayer.

The tinting is done by drawing a colored rect above the image with the `SourceIn` compositing mode which draws only above opaque content. 

Here the sublayer colored with the tint color and set up with the compositing filter plays that role. The SourceIn compositing of the image layer together with the tint colored sublayer set on top of the image is what results in the tinted output.

Reply:
Good call to double check on the layer-backing limits! It's nice to know this is being tracked on PRs.